README.md
# CoinGecko API TypeScript Wrapper

[![GitHub contributors](https://img.shields.io/github/contributors/michael-siek/coingecko-api)](https://github.com/michael-siek/coingecko-api/graphs/contributors)
[![npm](https://img.shields.io/npm/dt/@crypto-coffee/coingecko-api)](https://www.npmjs.com/package/@crypto-coffee/coingecko-api)

> An open-source TypeScript wrapper for the the CoinGecko API

[Coingecko API Documentation](https://www.coingecko.com/en/api/documentation)

## Installation

```console
npm install @crypto-coffee/coingecko-api
```

## Usage Example

```js
import CoinGeckoApi from '@crypto-coffee/coingecko-api'

;(async () => {
  try {
    const coinGeckoApi = new CoinGeckoApi()
    const results = await coinGeckoApi.simple({
      ids: 'Bitcoin',
      vs_currencies: 'usd'
    })
    console.log(results)
  } catch (err) {
    // do something with the error
  }
})()
```

## Endpoints

We have supplied the majority of CoinGecko's API. The available endpoints are supplied below:

### `Ping`

```js
const result = await coinGeckoApi.ping()
```

### `/simple/price`

```js
// Single id
const results = await coinGeckoApi.simple({
    vs_currencies: 'usd'
    id: 'bitcoin'
})
// Multiple ids
const results = await coinGeckoApi.simple({
    vs_currencies: 'usd'
    id: ['bitcoin', 'Ethereum']
})
```

### `/simple/token_price/{id}`

```js
// single contract_address
const results = await coinGeckoApi.simpleTokenPrice({
    'binance-smart-chain', {
    contract_addresses: '0x00',
    vs_currencies: 'usd'
    }
})
// Multiple contract_addresses
const results = await coinGeckoApi.simpleTokenPrice({
    'binance-smart-chain', {
    contract_addresses: ['0x00', '0x001'],
    vs_currencies: 'usd'
    }
})
```

### `/simple/supported_vs_currencies`

```js
const results = await coinGeckoApi.supportedCurrencies()
```

### `​/coins​/list`

```js
const results = await coinGeckoApi.coinList()
```

### `​/coins​/markets`

```js
const results = await coinGeckoApi.coinMarkets({ vs_currency: 'usd' })
```

### `​/coins​/{id}`

```js
const results = await coinGeckoApi.coinMarkets({ id: 'bitcoin' })
```

### `​/coins​/{id}​/tickers`

```js
const results = await coinGeckoApi.coinTickers('bitcoin', {
  exchange_ids: 'binance'
})

const { data, request } = await coinGeckoApi.coinTickers('bitcoin', {
  exchange_ids: 'binance'
})
```

### `​/coins​/{id}​/history`

```js
const results = await coinGeckoApi.coinHistory('bitcoin', '30-12-2017', {
  localization: false
})
const results = await coinGeckoApi.coinHistory('bitcoin', '30-12-2017')
```

### `​/coins​/{id}​/market_chart`

```js
const results = await coinGeckoApi.coinMarketChart('bitcoin', {
  vs_currency: 'usd',
  days: '1'
})
```

### `​/coins​/{id}​/market_chart​/range`

```js
const results = await coinGeckoApi.coinMarketChartRange('bitcoin', {
  vs_currency: 'usd',
  from: '1392577232',
  to: '1422577232'
})
```

### `​/coins​/{id}​/ohlc`

```js
const results = await coinGeckoApi.coinOHLC('bitcoin', {
  vs_currency: 'usd',
  days: 7
})
```

​

### `/coins​/{id}​/contract​/{contract_address}`

```js
const results = await coinGeckoApi.contractInformation(
  'binance-smart-chain',
  '0x00'
)
```

### `/coins​/{id}​/contract​/{contract_address}​/market_chart​/`

```js
const results = await coinGeckoApi.contractMarketChart(
  'binance-smart-chain',
  '0x00',
  {
    vs_currency: 'usd',
    days: '1'
  }
)
```

### `/coins​/{id}​/contract​/{contract_address}​/market_chart​/range`

```js
const results = await coinGeckoApi.contractMarketChartRange(
  'binance-smart-chain',
  '0x00',
  {
    vs_currency: 'usd',
    from: '1619881896',
    to: '1619968296'
  }
)
```

### `/asset_platforms`

```js
const results = await coinGeckoApi.assetPlatforms()
```

### `/coins​/categories​/list`

```js
const results = await coinGeckoApi.categoriesList()
```

### `/coins​/categories`

```js
const results = await coinGeckoApi.categoriesListMarketData()
```

### `/exchanges`

```js
const results = await coinGeckoApi.exchangesList()
```

### `/exchanges​/list`

```js
const results = await coinGeckoApi.exchangesList()
```

### `/exchanges​/{id}`

```js
const results = await coinGeckoApi.exchangesById('binance')
```

### `/exchanges​/{id}​/tickers`

```js
const results = await coinGeckoApi.exchangeTickers('binance')
```

### `/exchanges​/{id}​/volume_chart`

```js
const results = await coinGeckoApi.exchangesVolumeChart('binance', 1)
```

### Contributions

This package is maintained by [Zidious](https://github.com/Zidious)(Gabe) and [Michael](https://github.com/michael-siek). If any of the endpoints that are supplied by CoinGecko and are not in this package please only an issue and we will get them added as soon as possible.

:coffee: :coffee:


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.2.0 (2022-10-10)


### Features

* add new `companies/public_treasury/{coin_id} ` endpoint ([#29](https://github.com/michael-siek/coingecko-api/issues/29)) ([a7b3b0c](https://github.com/michael-siek/coingecko-api/commit/a7b3b0c6ff92968ae9e036909aff47d713e2c68f))
* create coingecko api library ([#1](https://github.com/michael-siek/coingecko-api/issues/1)) ([5dc89c5](https://github.com/michael-siek/coingecko-api/commit/5dc89c57e04f912bda19f159e2e9399f2ffac043))
* typed API responses ([#37](https://github.com/michael-siek/coingecko-api/issues/37)) ([a0d5a68](https://github.com/michael-siek/coingecko-api/commit/a0d5a68a100a81e6c7de8d55631689e3228f22d1))


### Bug Fixes

* allow full list of `vs_currencies` to be consumed ([#39](https://github.com/michael-siek/coingecko-api/issues/39)) ([0fcc675](https://github.com/michael-siek/coingecko-api/commit/0fcc6758315583e676d9faf21a146bd3ae63f3d5))
* only package required files ([#22](https://github.com/michael-siek/coingecko-api/issues/22)) ([e8ee5ee](https://github.com/michael-siek/coingecko-api/commit/e8ee5eeda223aa59ceeddf09bfebcdba54688330))
* typings ([#20](https://github.com/michael-siek/coingecko-api/issues/20)) ([0717f59](https://github.com/michael-siek/coingecko-api/commit/0717f59a3fa94f5d2c159dfba6b5aeda2bc31b67))

## 1.1.0 (2022-04-04)


### Features

* add new `companies/public_treasury/{coin_id} ` endpoint ([#29](https://github.com/michael-siek/coingecko-api/issues/29)) ([a7b3b0c](https://github.com/michael-siek/coingecko-api/commit/a7b3b0c6ff92968ae9e036909aff47d713e2c68f))
* create coingecko api library ([#1](https://github.com/michael-siek/coingecko-api/issues/1)) ([5dc89c5](https://github.com/michael-siek/coingecko-api/commit/5dc89c57e04f912bda19f159e2e9399f2ffac043))


### Bug Fixes

* only package required files ([#22](https://github.com/michael-siek/coingecko-api/issues/22)) ([e8ee5ee](https://github.com/michael-siek/coingecko-api/commit/e8ee5eeda223aa59ceeddf09bfebcdba54688330))
* typings ([#20](https://github.com/michael-siek/coingecko-api/issues/20)) ([0717f59](https://github.com/michael-siek/coingecko-api/commit/0717f59a3fa94f5d2c159dfba6b5aeda2bc31b67))


src/test.ts
import CoinGeckoAPI from '.'

const gecko = new CoinGeckoAPI()

;(async () => {
  const data = await gecko.coins('bitcoin')

  console.log(data)
})()


src/tests/test.ts
import 'mocha'
import { assert } from 'chai'
import CoinGeckoAPI from '..'

describe('CoinGeckoAPI', () => {
  const coinGeckoApi = new CoinGeckoAPI()

  describe('.ping', () => {
    it('Should ping server', async () => {
      const result = await coinGeckoApi.ping()
      assert.isObject(result)
    })
  })

  describe('.simple', () => {
    it('Returns the price of BTC when passed as a string', async () => {
      const data = await coinGeckoApi.simple({
        ids: 'Bitcoin',
        vs_currencies: 'usd'
      })

      assert.isObject(data)
      assert.isNotNull(data.bitcoin)
      assert.isUndefined(data.ethereum)
    })

    it('Returns the price of BTC/ETH when passed as an array', async () => {
      const data = await coinGeckoApi.simple({
        ids: ['Bitcoin', 'Ethereum'],
        vs_currencies: 'usd'
      })

      assert.isObject(data)
      assert.isNotNull(data.bitcoin)
      assert.isNotNull(data.ethereum)
    })
  })

  describe('.simple/token_price', () => {
    it('Returns the token address (CAKE) and the price', async () => {
      const data = await coinGeckoApi.simpleTokenPrice('binance-smart-chain', {
        contract_addresses: '0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82',
        vs_currencies: 'usd'
      })

      assert.isObject(data)
      assert.isNotNull(data[0])
    })

    it('Returns the price of CAKE and BUNNY when passed as an array', async () => {
      const data = await coinGeckoApi.simpleTokenPrice('binance-smart-chain', {
        contract_addresses: [
          '0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82',
          '0xc9849e6fdb743d08faee3e34dd2d1bc69ea11a51'
        ],
        vs_currencies: 'usd'
      })

      assert.isObject(data)
      assert.isNotNull(data[0])
      assert.isNotNull(data[1])
    })
  })

  describe('./simple/supported_vs_currencies', () => {
    it('Return an array of supported currencies', async () => {
      const data = await coinGeckoApi.supportedCurrencies()

      assert.isArray(data)
      assert.isNotNull(data[0])
    })
  })

  describe('.coins/list', () => {
    it('Return list of all the supported coins', async () => {
      const data = await coinGeckoApi.coinList()
      const { id, symbol, name } = data[0]
      assert.isArray(data)
      assert.isNotNull(id)
      assert.isNotNull(symbol)
      assert.isNotNull(name)
    })
  })

  describe('.coins/market', () => {
    it('Return all the coins market data (price, market cap, volume)', async () => {
      const data = await coinGeckoApi.coinMarkets({
        vs_currency: 'usd'
      })
      const { id, symbol, name } = data[0]

      assert.isArray(data)
      assert.isNotNull(id)
      assert.isNotNull(symbol)
      assert.isNotNull(name)
    })
  })

  describe('.coins/{id}', () => {
    it('Returns current data (name, price, market, ... including exchange tickers) for a coin', async () => {
      const data = await coinGeckoApi.coins('bitcoin')

      assert.isObject(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.symbol)
      assert.isNotNull(data.name)
    })
  })

  describe('.coins/{id}/tickers', () => {
    it('Returns coin tickers (paginated to 100 items)', async () => {
      const data = await coinGeckoApi.coinTickers('bitcoin')

      assert.isObject(data)
      assert.isNotNull(data.name)
    })
  })

  describe('.coins/{id}/history', () => {
    it('Returns historical data (name, price, market, stats) at a given date for a coin with localization set to false', async () => {
      const data = await coinGeckoApi.coinHistory('bitcoin', '30-12-2017', {
        localization: false
      })

      assert.isObject(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.symbol)
      assert.isNotNull(data.name)
    })

    it('Returns historical data (name, price, market, stats) at a given date for a coin', async () => {
      const data = await coinGeckoApi.coinHistory('bitcoin', '30-12-2017')

      assert.isObject(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.symbol)
      assert.isNotNull(data.name)
    })
  })

  describe('.coins/{id}/market_chart', () => {
    it('Return historical market data include price, market cap, and 24h volume (granularity auto)', async () => {
      const data = await coinGeckoApi.coinMarketChart('bitcoin', {
        vs_currency: 'usd',
        days: '1'
      })

      assert.isObject(data)
      assert.isNotNull(data.prices)
      assert.isNotNull(data.market_caps)
      assert.isNotNull(data.total_volumes)
    })
  })

  describe('.coins/{id}/market_chart/range', () => {
    it('Return historical market data include price, market cap, and 24h volume within a range of timestamp (granularity auto)', async () => {
      const data = await coinGeckoApi.coinMarketChartRange('bitcoin', {
        vs_currency: 'usd',
        from: '1392577232',
        to: '1422577232'
      })

      assert.isObject(data)
      assert.isNotNull(data.prices)
      assert.isNotNull(data.market_caps)
      assert.isNotNull(data.total_volumes)
    })
  })

  describe('.coins/{id}/ohlc', () => {
    it('Return candles body: 1-2 days: 30mins // 3-30 days: 4hours // 31 and before: 4 days', async () => {
      const data = await coinGeckoApi.coinOHLC('bitcoin', {
        vs_currency: 'usd',
        days: 7
      })

      assert.isArray(data)
    })
  })

  describe('.coins/{id}/contract/{contract_address}', () => {
    it('Return coin info from contract address', async () => {
      const data = await coinGeckoApi.contractInformation(
        'binance-smart-chain',
        '0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82'
      )

      assert.isObject(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.symbol)
      assert.isNotNull(data.name)
    })
  })

  describe('.coins/{id}/contract/{contract_address}/market_chart', () => {
    it('Return historical market data include price, market cap, and 24h volume (granularity auto)', async () => {
      const data = await coinGeckoApi.contractMarketChart(
        'binance-smart-chain',
        '0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82',
        {
          vs_currency: 'usd',
          days: '1'
        }
      )

      assert.isObject(data)
      assert.isNotNull(data.prices)
      assert.isNotNull(data.market_caps)
      assert.isNotNull(data.total_volumes)
    })
  })

  describe('.coins/{id}/contract/{contract_address}/market_chart/range', () => {
    it('Return historical market data include price, market cap, and 24h volume within a range of timestamp (granularity auto)', async () => {
      const data = await coinGeckoApi.contractMarketChartRange(
        'binance-smart-chain',
        '0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82',
        {
          vs_currency: 'usd',
          from: '1619881896',
          to: '1619968296'
        }
      )

      assert.isObject(data)
      assert.isNotNull(data.prices)
      assert.isNotNull(data.market_caps)
      assert.isNotNull(data.total_volumes)
    })
  })

  describe('.asset_platforms', () => {
    it('Return list of all asset platforms', async () => {
      const data = await coinGeckoApi.assetPlatforms()

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.chain_identifier)
      assert.isNotNull(data.name)
    })
  })

  describe('./coins/categories/list', () => {
    it('Return list of all categories', async () => {
      const data = await coinGeckoApi.categoriesList()

      assert.isArray(data)
      assert.isNotNull(data.category_id)
      assert.isNotNull(data.name)
    })
  })

  describe('./coins/categories', () => {
    it('Return list of all categories with market data (market_cap_desc (default))', async () => {
      const data = await coinGeckoApi.categoriesListMarketData()

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
      assert.isNotNull(data.market_cap)
      assert.isNotNull(data.market_cap_change_24h)
      assert.isNotNull(data.volume_24h)
      assert.isNotNull(data.updated_at)
    })
    it('Return list of all categories with market data (market_cap_asc)', async () => {
      const data = await coinGeckoApi.categoriesListMarketData({
        order: 'market_cap_asc'
      })

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
      assert.isNotNull(data.market_cap)
      assert.isNotNull(data.market_cap_change_24h)
      assert.isNotNull(data.volume_24h)
      assert.isNotNull(data.updated_at)
    })
  })

  describe('./exchanges', () => {
    it('Return list of all exchanges', async () => {
      const data = await coinGeckoApi.exchangesList()

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
      assert.isNotNull(data.year_established)
      assert.isNotNull(data.country)
      assert.isNotNull(data.description)
      assert.isNotNull(data.url)
      assert.isNotNull(data.image)
      assert.isNotNull(data.has_trading_incentive)
      assert.isNotNull(data.trust_score)
      assert.isNotNull(data.trust_score_rank)
      assert.isNotNull(data.trade_volume_24h_btc)
      assert.isNotNull(data.trade_volume_24h_btc_normalized)
    })

    it('Return list of all exchanges (page limit)', async () => {
      const data = await coinGeckoApi.exchanges({
        per_page: 1,
        page: '1'
      })

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
      assert.isNotNull(data.year_established)
      assert.isNotNull(data.country)
      assert.isNotNull(data.description)
      assert.isNotNull(data.url)
      assert.isNotNull(data.image)
      assert.isNotNull(data.has_trading_incentive)
      assert.isNotNull(data.trust_score)
      assert.isNotNull(data.trust_score_rank)
      assert.isNotNull(data.trade_volume_24h_btc)
      assert.isNotNull(data.trade_volume_24h_btc_normalized)
    })
  })

  describe('./exchanges/list', () => {
    it('Return all supported markets id and name', async () => {
      const data = await coinGeckoApi.exchangesList()

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
    })
  })

  describe('./exchanges/{id}', () => {
    it('Return exchange volume in BTC and tickers', async () => {
      const data = await coinGeckoApi.exchangesById('binance')

      assert.isObject(data)
      assert.isNotNull(data.name)
      assert.isNotNull(data.year_established)
      assert.isNotNull(data.country)
    })
  })

  describe('./exchanges/{id}/tickers', () => {
    it('Return exchange tickers', async () => {
      const data = await coinGeckoApi.exchangeTickers('binance')

      assert.isObject(data)
      assert.isNotNull(data.name)
      assert.isNotNull(data.tickers)
    })
  })

  describe('./exchanges/{id}/volume_chart', () => {
    it('Return volume chart data for a given exchange', async () => {
      const data = await coinGeckoApi.exchangesVolumeChart('binance', 1)

      assert.isArray(data)
      assert.isNotNull(data)
    })
  })

  describe('./indexes', () => {
    it('Return list of all market indexes', async () => {
      const data = await coinGeckoApi.indexes()

      assert.isArray(data)
      assert.isNotNull(data.name)
      assert.isNotNull(data.id)
      assert.isNotNull(data.market)
    })
  })

  describe('./indexes/list', () => {
    it('Return list of all market indexes', async () => {
      const data = await coinGeckoApi.indexesList()

      assert.isArray(data)
      assert.isNotNull(data.id)
      assert.isNotNull(data.name)
    })
  })

  describe('./companies/public_treasury/{coin_id}', () => {
    it('Return list of companies holding bitcoin or ethereum', async () => {
      const data = await coinGeckoApi.companies('ethereum')

      assert.isObject(data)
      assert.isNotNull(data.name)
      assert.isNotNull(data.symbol)
      assert.isArray(data.companies)
      assert.isNotNull(data.companies[0].total_holdings)
    })
  })
})


src/types.d.ts
// Simple Price Base Interface
interface SimplePriceBaseParams {
  vs_currencies: 'usd'
  include_market_cap?: boolean
  include_24hr_vol?: boolean
  include_24hr_change?: boolean
  include_last_updated_at?: boolean
}

interface MarketChartBaseParams {
  // (required) The target currency of market data (usd, eur, jpy, etc.)
  vs_currency: 'usd'
  // (required) Data up to number of days ago (eg. 1,14,30,max)
  days: string
}

interface PageBaseParams {
  per_page?: string | number
  page?: string | number
}

export interface SimplePriceParams extends SimplePriceBaseParams {
  // (required) can be singular {ids: 'bitcoin'} or {ids: ['Bitcoin','Ethereum']}
  ids: string | string[]
}

export interface SimpleTokenPriceParams extends SimplePriceBaseParams {
  // (required) can be singular {contract_addresses: '0x000'} or {contract_addresses: ['0x000','0x001']}
  contract_addresses: string | string[]
}

export interface CoinListParams {
  include_platform?: boolean
}

export interface CoinMarketParams extends PageBaseParams {
  // (required) The target currency of market data (usd, eur, jpy, etc.)
  vs_currency: 'usd'
  ids?: string
  category?: string
  order?: string
  sparkline?: boolean
  price_change_percentage?: string
}

export interface CoinsParams {
  localization?: string
  tickers?: boolean
  market_data?: boolean
  community_data?: boolean
  developer_data?: boolean
  sparkline?: boolean
}

export interface CoinsTickersParams {
  exchange_ids?: string
  include_exchange_logo?: string
  page?: number
  order?: string
  depth?: string
}

export interface CoinsHistoryParams {
  localization?: boolean
}

export interface CoinsMarketChartParams extends MarketChartBaseParams {
  interval?: string
}

export interface MarketChartRangeBaseParams {
  // (required) The target currency of market data (usd, eur, jpy, etc.)
  vs_currency: 'usd'
  // (required) From date in UNIX Timestamp (eg. 1392577232)
  from: string
  // (required) To date in UNIX Timestamp (eg. 1422577232)
  to: string
}

export interface CoinOhlcParams {
  // (required) The target currency of market data (usd, eur, jpy, etc.)
  vs_currency: 'usd'
  // (required) Data up to number of days ago (1/7/14/30/90/180/365/max)
  days: number
}

export interface CategoriesOrderParams {
  order?:
    | 'market_cap_asc'
    | 'name_desc'
    | 'name_asc'
    | 'market_cap_change_24h_desc'
    | 'market_cap_change_24h_asc'
}

export interface ExchangesTickersParams {
  coin_ids?: string
  include_exchange_logo?: string
  page?: number
  depth?: string
  order?: string
}

export interface FinanceProductParams extends PageBaseParams {
  start_at?: string
  end_at?: string
}


src/index.ts
import assert from 'assert'
import type { AxiosInstance } from 'axios'
import * as querystring from 'query-string'
import { API_CONNECTOR } from './utils'
import type {
  CoinListParams,
  CoinMarketParams,
  CoinsParams,
  CoinsTickersParams,
  SimplePriceParams,
  CoinsHistoryParams,
  SimpleTokenPriceParams,
  CoinsMarketChartParams,
  MarketChartRangeBaseParams,
  PageBaseParams,
  CoinOhlcParams,
  MarketChartBaseParams,
  CategoriesOrderParams,
  ExchangesTickersParams,
  FinanceProductParams,
  CompaniesHoldings,
  Ping,
  Simple,
  SimpleTokenPrice,
  CoinList,
  CoinMarkets,
  Coins,
  Tickers,
  CoinHistory,
  CoinMarketChart,
  ContractInformation
} from './types'

export default class CoinGeckoAPI {
  private axios: AxiosInstance

  constructor() {
    this.axios = API_CONNECTOR
  }

  public async ping(): Promise<Ping> {
    const method = 'ping'
    return await this.get(method)
  }

  // Simple Endpoints

  /**
   *
   * @param {SimplePriceParams} params - Object to pass through
   */

  public async simple(params: SimplePriceParams): Promise<Simple> {
    const method = 'simple/price'
    return await this.get(method, params)
  }

  /**
   * @param id - (required) The ID of the platform to fetch
   * @param {SimpleTokenPriceParams} params - Object to pass through
   */

  public async simpleTokenPrice(
    id: string,
    params: SimpleTokenPriceParams
  ): Promise<SimpleTokenPrice> {
    const method = 'simple/token_price/' + id
    return await this.get(method, params)
  }

  public async supportedCurrencies(): Promise<string[]> {
    const method = 'simple/supported_vs_currencies'
    return await this.get(method)
  }

  // Coin Endpoints

  /**
   *
   * @param {CoinListParams} params - Object to pass through
   */

  public async coinList(params?: CoinListParams): Promise<CoinList[]> {
    const method = 'coins/list'
    return await this.get(method, params)
  }

  /**
   *
   * @param {CoinMarketParams} params - Object to pass through
   */

  public async coinMarkets(params?: CoinMarketParams): Promise<CoinMarkets[]> {
    const method = 'coins/markets'
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param {CoinsParams} params - Object to pass through
   */

  public async coins(id: string, params?: CoinsParams): Promise<Coins> {
    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param {CoinsTickersParams} params - Object to pass through
   */

  public async coinTickers(
    id: string,
    params?: CoinsTickersParams
  ): Promise<Tickers> {
    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id + '/tickers'
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param date (required) Pass data of data snapshot dd-mm-yyyy e.g. 30-12-2017
   * @param {CoinsHistoryParams} params - Object to pass through
   */

  public async coinHistory(
    id: string,
    date: string,
    params?: CoinsHistoryParams
  ): Promise<CoinHistory> {
    assert(
      date,
      'Date must conform to this standard: dd-mm-yyyy e.g. 30-12-2017'
    )

    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id + '/history?date=' + date
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param {CoinsMarketChartParams} params - Object to pass through
   */

  public async coinMarketChart(
    id: string,
    params: CoinsMarketChartParams
  ): Promise<CoinMarketChart> {
    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id + '/market_chart'
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param {MarketChartRangeBaseParams} params - Object to pass through
   */

  public async coinMarketChartRange(
    id: string,
    params: MarketChartRangeBaseParams
  ): Promise<CoinMarketChart> {
    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id + '/market_chart/range'
    return await this.get(method, params)
  }

  /**
   * @param id (required) Pass the coin id e.g. bitcoin.
   * @param {CoinOhlcParams} params - Object to pass through
   */

  public async coinOHLC(
    id: string,
    params: CoinOhlcParams
  ): Promise<number[][]> {
    assert(id, 'The ID of the coin is required e.g. Bitcoin.')

    const method = 'coins/' + id + '/ohlc'
    return await this.get(method, params)
  }

  // Contract Endpoints

  /**
   * @param id (required) Asset Platform e.g. binance-smart-chain
   * @param contract_address - Token's Contract Address
   */

  public async contractInformation(
    id: string,
    contract_address: string
  ): Promise<ContractInformation> {
    assert(id, 'The ID of said asset platform e.g. binance-smart-chain')

    const method = 'coins/' + id + '/contract/' + contract_address
    return await this.get(method)
  }

  /**
   * @param id (required) Asset Platform e.g. binance-smart-chain
   * @param contract_address - Token's Contract Address
   * @param {MarketChartBaseParams} - Object to pass through
   */

  public async contractMarketChart(
    id: string,
    contract_address: string,
    params: MarketChartBaseParams
  ): Promise<CoinMarketChart> {
    assert(id, 'The ID of said asset platform e.g. binance-smart-chain.')
    assert(contract_address, 'Token contract address required.')

    const method =
      'coins/' + id + '/contract/' + contract_address + '/market_chart'

    return await this.get(method, params)
  }

  /**
   * @param id (required) Asset Platform e.g. binance-smart-chain
   * @param contract_address - Token's Contract Address
   * @param {MarketChartRangeBaseParams} - Object to pass through
   */

  public async contractMarketChartRange(
    id: string,
    contract_address: string,
    params: MarketChartRangeBaseParams
  ): Promise<CoinMarketChart> {
    assert(id, 'The ID of said asset platform e.g. binance-smart-chain.')
    assert(contract_address, 'Token contract address required.')

    const method =
      'coins/' + id + '/contract/' + contract_address + '/market_chart/range'

    return await this.get(method, params)
  }

  // Asset Platforms Endpoint

  public async assetPlatforms() {
    const method = 'asset_platforms'
    return await this.get(method)
  }

  // Categories Endpoints

  public async categoriesList() {
    const method = 'coins/categories/list'
    return await this.get(method)
  }

  /**
   * @param {CategoriesOrderParams} - Object to pass through
   */

  public async categoriesListMarketData(params?: CategoriesOrderParams) {
    const method = 'coins/categories'
    return await this.get(method, params)
  }

  // Exchanges Endpoints

  /**
   * @param {PageBaseParams} - Object to pass through
   */

  public async exchanges(params?: PageBaseParams) {
    const method = 'exchanges'
    return await this.get(method, params)
  }

  public async exchangesList() {
    const method = 'exchanges/list'
    return await this.get(method)
  }

  /**
   * @param id (required) ID of the exchange e.g. binance
   */

  public async exchangesById(id: string) {
    assert(id, 'Pass the exchange ID e.g. binance')

    const method = 'exchanges/' + id
    return await this.get(method)
  }

  /**
   * @param id (required) ID of the exchange e.g. binance
   * @param {ExchangesTickersParams} - Object to pass through
   */

  public async exchangeTickers(id: string, params?: ExchangesTickersParams) {
    assert(id, 'Pass the exchange ID e.g. binance')

    const method = 'exchanges/' + id + '/tickers'
    return await this.get(method, params)
  }

  /**
   * @param id (required) ID of the exchange e.g. binance
   * @param days (required) Data up to number of days ago (eg. 1,14,30)
   */

  public async exchangesVolumeChart(id: string, days: number) {
    assert(id, 'Pass the exchange ID e.g. binance')

    const method = 'exchanges/' + id + '/volume_chart?days=' + days
    return await this.get(method)
  }

  // Finance Endpoints

  /**
   * @param {FinanceProductParams} params - Object to pass in
   */

  public async financeProducts(params?: FinanceProductParams) {
    const method = 'finance_products'
    return await this.get(method, params)
  }

  // Indexes Endpoints

  /**
   * @param {PageBaseParams} params - Object to pass in
   */

  public async indexes(params?: PageBaseParams) {
    const method = 'indexes'
    return await this.get(method, params)
  }

  /**
   * @param market_id - Pass the market id (can be obtained from ./exchanges/list)
   * @param id - Pass the index id (can be obtained from ./indexes/list)
   */

  public async indexesByMarketIndexId(market_id: string, id: string) {
    assert(id, 'Pass the index id (can be obtained from /indexes/list)')
    assert(
      market_id,
      'Pass the market id (can be obtained from ./exchanges/list)'
    )

    const method = 'indexes/' + market_id + '/' + id
    return await this.get(method)
  }

  public async indexesList() {
    const method = 'indexes/list'
    return await this.get(method)
  }

  /**
   *
   * @param {CompaniesHoldings} coin_id - bitcoin or ethereum
   */
  public async companies(coin_id: CompaniesHoldings) {
    const method = 'companies/public_treasury/' + coin_id
    return await this.get(method)
  }

  // Get and path builder functions

  private async get(method: string, params?: any) {
    const endpoint = this.build_request_path(method, params)
    const { data } = await this.axios.get(endpoint)
    return data
  }

  private build_request_path(path: string, params?: any) {
    let queryParams = ''
    if (Array.isArray(params?.ids) && params?.ids) {
      params.ids = params.ids.join(',')
    }

    if (
      Array.isArray(params?.contract_addresses) &&
      params?.contract_addresses
    ) {
      params.contract_addresses = params.contract_addresses.join(',')
    }

    if (typeof params === 'object') {
      queryParams = querystring.stringify(params)
    }

    path = queryParams ? `/${path}?${queryParams}` : `/${path}`

    return path
  }
}
